File: src/super-board.ts
import { Board } from "./board";
import { Cell } from "./cell";

export class SuperBoard {
  private superCells: Board[];

  constructor(private boardSize: number) {
    this.superCells = new Array(boardSize * boardSize);

    for (let i = 0; i < this.superCells.length; i++) {
      this.superCells[i] = new Board(boardSize);
    }
  }

  setSuperCell(x: number, y: number, superCell: Board) {
    this.superCells[x + y * this.boardSize] = superCell;
  }

  getSuperCell(x: number, y: number) {
    return this.superCells[x + y * this.boardSize];
  }

  isRowFilled(y: number) {
    // a cell is filled if it is superCell.isDone();
    return this.superCells
      .slice(y * this.boardSize, y * this.boardSize + this.boardSize)
      .every((superCell) => superCell.isDone());
  }

  isColumnFilled(x: number) {
    return this.superCells
      .filter((_, i) => i % this.boardSize === x)
      .every((superCell) => superCell.isDone());
  }

  isDiagonalFilled() {
    return (
      this.superCells
        .filter((_, i) => i % (this.boardSize + 1) === 0)
        .every((superCell) => superCell.isDone()) ||
      this.superCells
        .filter((_, i) => i % (this.boardSize - 1) === this.boardSize - 1)
        .every((superCell) => superCell.isDone())
    );
  }

  isBoardFilled() {
    return this.superCells.every((superCell) => superCell.isDone());
  }

  getWinner() {
    for (let i = 0; i < this.boardSize; i++) {
      if (this.isRowFilled(i)) {
        return this.superCells[i * this.boardSize].getWinner();
      }
      if (this.isColumnFilled(i)) {
        return this.superCells[i].getWinner();
      }
    }

    if (this.isDiagonalFilled()) {
      return this.superCells[this.boardSize].getWinner();
    }

    return Cell.Empty;
  }

  isDraw() {
    return this.getWinner() === null && this.isBoardFilled();
  }

  isDone() {
    return this.isDraw() || this.getWinner() !== Cell.Empty;
  }
}

File: super-ttt-server/lib/super_ttt/logic/super_board.dart
import "package:super_ttt_server/super_ttt/logic/board.dart";
import "package:super_ttt_server/super_ttt/logic/cell.dart";

class SuperBoard {
  List<Board> superCells = [];
  int boardSize;

  SuperBoard({this.boardSize = 3}) {
    for (int i = 0; i < boardSize * boardSize; i++) {
      superCells.add(Board(boardSize: boardSize));
    }
  }

  setSuperCell(int x, int y, Board superCell) {
    superCells[x + y * boardSize] = superCell;
  }

  void disable() {
    for (int i = 0; i < boardSize * boardSize; i++) {
      superCells[i].setActive(false);
    }
  }

  Board getSuperCell(int x, int y) {
    return superCells[x + y * boardSize];
  }

  bool isRowFilled(int row, Cell player) {
    for (int i = 0; i < boardSize; i++) {
      if (getSuperCell(i, row).getWinner() != player) {
        return false;
      }
    }
    return true;
  }

  bool isColumnFilled(int column, Cell player) {
    for (int i = 0; i < boardSize; i++) {
      if (getSuperCell(column, i).getWinner() != player) {
        return false;
      }
    }
    return true;
  }

  bool isDiagonalFilled(Cell player) {
    for (int i = 0; i < boardSize; i++) {
      if (getSuperCell(i, i).getWinner() != player) {
        return false;
      }
    }
    return true;
  }

  bool isAntiDiagonalFilled(Cell player) {
    for (int i = 0; i < boardSize; i++) {
      if (getSuperCell(i, boardSize - i - 1).getWinner() != player) {
        return false;
      }
    }
    return true;
  }

  bool isDone() {
    for (int i = 0; i < boardSize; i++) {
      if (isRowFilled(i, Cell.x) ||
          isColumnFilled(i, Cell.x) ||
          isDiagonalFilled(Cell.x) ||
          isAntiDiagonalFilled(Cell.x)) {
        return true;
      }
      if (isRowFilled(i, Cell.o) ||
          isColumnFilled(i, Cell.o) ||
          isDiagonalFilled(Cell.o) ||
          isAntiDiagonalFilled(Cell.o)) {
        return true;
      }
    }
    return false;
  }

  Cell getWinner() {
    for (int i = 0; i < boardSize; i++) {
      if (isRowFilled(i, Cell.x) ||
          isColumnFilled(i, Cell.x) ||
          isDiagonalFilled(Cell.x) ||
          isAntiDiagonalFilled(Cell.x)) {
        return Cell.x;
      }
      if (isRowFilled(i, Cell.o) ||
          isColumnFilled(i, Cell.o) ||
          isDiagonalFilled(Cell.o) ||
          isAntiDiagonalFilled(Cell.o)) {
        return Cell.o;
      }
    }
    return Cell.empty;
  }

  bool isDraw() {
    return isDone() && getWinner() == Cell.empty;
  }

  Map<String, dynamic> toJson() {
    return {
      "superCells": superCells.map((e) => e.toJson()).toList(),
    };
  }
}
